typings/composer.d.ts

```diff
@@ -0,0 +1,197 @@
+/** @format */
+
+import * as tt from './telegram-types.d'
+import { ContextMessageUpdate } from './context'
+
+type HearsTriggers = string[] | string | RegExp | RegExp[] | Function
+
+export interface MiddlewareFn<TContext extends ContextMessageUpdate> {
+  /*
+  next's parameter is in a contravariant position, and thus, trying to type it
+  prevents assigning `MiddlewareFn<ContextMessageUpdate>`
+  to `MiddlewareFn<CustomContext>`.
+  Middleware passing the parameter should be a separate type instead.
+  */
+  (ctx: TContext, next: () => Promise<unknown>): unknown
```

Comment:
I am thinking about best practices here...

As a first idea I would say this should be kind of strict: `(ctx: TContext: next: () => Promise<void>): void | Promise<void>`.
But its neat to do quick one liners like this:
```js
bot.command('start', ctx => ctx.reply('go away'))
```

In that case `ctx.reply` returns `Promise<Message>` which can not be applied to `Promise<void>`.
So `Promise<unknown>` would be better for these one liners.
I currently do not see any point for something returned which isnt a Promise. So `void | Promise<unknown>` as a return type seems quite good for me.

Looking at next: next should not return something? Yeah, its possible and `next` can return something but that feels hacky and not intended. ([If someone wants to get stuff from one middleware to the next they should use `ctx.state`.](https://telegraf.js.org/#/?id=state)) So I think `() => Promise<void>` would represent the purpose of `next` better?

I do not really understand the comment you added about the next parameter. It seems unrelated to the next parameter? The next parameter is not changing when the TContext is a different one.

Based on that I would suggest this:

```suggestion
  (ctx: TContext, next: () => Promise<void>): void | Promise<unknown>
```

Reply:
[`void` is weird](https://github.com/microsoft/TypeScript/issues/20006#issuecomment-344410732); I don't understand it and thus I don't trust it enough to write `Promise<void>`.

If you want stricter type, why allow returning `void` from `MiddlewareFn`? Why not just `Promise<unknown>`?

Reply:
First void: Its the absence of any type at all. Return type -> You except something to return. void as return type -> you expect the absence of anything. Which sounds more fitting to me than unknown: there is something but i don't know what it is. next shouldnt return anything. Technically it can when Middlewares are badly implemented but its not planned that way.

Second returning void.  When doing stuff locally or when doing long tasks in parallel for example:
```ts
bot.action('whatever', ctx => {
  ctx.session.lastAction = 'whatever'
})

bot.action('whatever', ctx => {
  setTimeout(() => ctx.reply('hey you!'), 1000)
})
```

Reply:
I see. Still I'm not convinced if the benefits outweight the temptation to just not `await`:

```ts
ctx => {
  const x = "result of some calculation"
  ctx.reply(x) // no await, no return, error can't be caught
}
```

`state` is currently untyped and I have no idea how it could ever be well-typed; I'm not sure if that's the best way for middlewares in TypeScript to communicate.

Reply:
You can use linters like typescript-eslint with [rules preventing floating promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md)

Also something like .catch can be used to prevent unhandled rejections:
```ts
ctx => {
  const x = 'result'
  ctx.reply(x).catch(error => console.error(error))
}
```

I use this when debouncing many user messages.
The user sends a bunch of messages which get saved and then replied to when there was no new message for some time.

Another use case is my chat record bot: Look at each message in a group and save it. It does not respond on any message. When the user uses /finish after an hour or so the whole chat log is written into a file and sent to the chat. (In this case saving is done asynchron so I want to await but this can be done with writeFileSync or something like that too.)

Reply:
The type of state is based on middlewares unrelated to telegraf itself. Telegraf itself does not use it so it shouldnt type it.

I would expect something like this on the user side:

```ts
import {Context as TelegrafContext} from 'telegraf'
interface MyContext extends TelegrafContext {
  state: {
    userSeemsFine: boolean;
  }
}

const bot = new Telegraf<MyContext>(token)
bot.use(async (ctx, next) => {
  ctx.state = {
    userSeemsFine = Math.random() > 0.5
  }
  return next()
})
bot.command('whatever', async ctx => ctx.reply(ctx.state.userSeemsFine ? 'yaay' : 'oh no'))
```

Personally I prefer to keep things externally as it does not require the context of the specific Telegram update so its something unrelated to the Context -> keep it separate.

```ts
const userSeemsFine: Map<number, bool> = new Map()

bot.use(async (ctx, next) => {
  if (ctx.from) {
    userSeemsFine[ctx.from.id] = Math.random() > 0.5
  }

  return next()
})
bot.command('whatever', async ctx => ctx.reply(userSeemsFine[ctx.from!.id] ? 'yaay' : 'oh no'))
```

Reply:
> I do not really understand the comment you added about the _`next`_ parameter. [...] The _`next`_ parameter is not changing when the TContext is a different one.

`next`_'s_ parameter.